source/slang/syntax.cpp

```diff
@@ -2509,7 +2509,160 @@ void Type::accept(IValVisitor* visitor, void* extra)
         return substValue;
     }
 
+    //
+    // TaggedUnionType
+    //
+
+    String TaggedUnionType::ToString()
+    {
+        String result;
+        result.append("__TaggedUnion(");
+        bool first = true;
+        for( auto caseType : caseTypes )
+        {
+            if(!first) result.append(", ");
+            first = false;
+
+            result.append(caseType->ToString());
+        }
+        result.append(")");
+        return result;
+    }
+
+    bool TaggedUnionType::EqualsImpl(Type* type)
+    {
+        auto taggedUnion = type->As<TaggedUnionType>();
+        if(!taggedUnion)
+            return false;
+
+        auto caseCount = caseTypes.Count();
+        if(caseCount != taggedUnion->caseTypes.Count())
+            return false;
+
+        for( UInt ii = 0; ii < caseCount; ++ii )
+        {
+            if(!caseTypes[ii]->Equals(taggedUnion->caseTypes[ii]))
+                return false;
+        }
+        return true;
+    }
+
+    int TaggedUnionType::GetHashCode()
+    {
+        int hashCode = 0;
+        for( auto caseType : caseTypes )
+        {
+            hashCode = combineHash(hashCode, caseType->GetHashCode());
+        }
+        return hashCode;
+    }
+
+    RefPtr<Type> TaggedUnionType::CreateCanonicalType()
```

Comment:
Yes, that would work, so we could add that as a small optimization.

Really, though, we need to revamp our type handling so that we are caching and re-using existing types when they have identical operands, rather than creating new `Type` objects all the time. Under that model, the overhead of creating a new type in this case would mostly be replaced with the overhead of a dictionary lookup based on the canonical case types.